### PR TITLE
Populate tls.Config.NextProtos for teleport specific ALPNs

### LINKED ISSUE #41500: Most Teleport-specific ALPN protocols don't confirm the selected ALPN protocol
Expected behavior:

Connecting to the Teleport web listener with a Teleport-specific ALPN (say, `teleport-proxy-ssh-grpc`) should result in the TLS ServerHello containing the negotiated protocol, as that will allow clients to not rely on potentially outdated autodetected state with regards to the presence of TLS terminators between the client and the control plane (that would trigger the HTTP Upgrade-based connection tunneling), and it will allow both parties to actually negotiate protocols among multiple ones that might be supported instead of just assuming that the chosen protocol is mutually supported.

Current behavior:

From a cursory look, the only protocol that works correctly is `teleport-reversetunnel`, with other protocols resulting in no protocol selection (which is hard to distinguish from a middlebox doing TLS interception with no idea of what to do with the Teleport-specific protocols); dishonorable mention to the magic `teleport-auth@<hex cluster name>` "tag" which only works on first position and completely changes the meaning of the ALPN selection, with the auth server actually requiring one of `h2` or `http/1.1` to be selected.

Bug details:
- Teleport version: likely since the introduction of ALPN for protocol selection
- Recreation steps: `openssl s_client -connect <proxyhostname>:443 -alpn teleport-proxy-ssh`

### PR DESCRIPTION
**Pull request by rosstimothy**
Any alpnproxy.HandlerDesc that forwards TLS or provides its own tls.Config is responsible for setting the tls.Config.NextProtos so that the negotiated ALPN is returned to clients. In all other cases the proxy web tls.Config is used which populates the next protos with _all_ supported ALPN protocols. All of the existing places that the default tls.Config was being overridden were supplying the NextProtos. This updates them to populate the NextProtos with the correct values and updates the commentary on alpnproxy.HandlerDecs to indicate that NextProtos should be set if not using the default tls.Config.

Fixes #41500.

Changelog: Ensure the negotiated ALPN is returned correctly for connections to Teleport.

### CODE REVIEW

File: lib/service/service.go [hunk 1]
```diff
@@ -4542,6 +4542,7 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 	}
 
 	tlscfg := serverTLSConfig.Clone()
+	tlscfg.NextProtos = []string{string(alpncommon.ProtocolHTTP2), string(alpncommon.ProtocolProxySSHGRPC)}
```
Comment (espadolini) [inline lib/service/service.go:4545]: Are we deliberately preferring `h2` to the custom Teleport next proto?

**Review by espadolini — APPROVED** [review]
Can we add a check for `ConnectionState().NegotiatedProtocol` in tests?

File: lib/service/service.go [hunk 2]
```diff
@@ -4542,6 +4542,7 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 	}
 
 	tlscfg := serverTLSConfig.Clone()
```
Comment (ibeckermayer) [inline lib/service/service.go:4544]: I notice this `serverTLSConfig` is being `Clone`-ed/used in multiple places in `initAuthService`, however its unclear why we prefer that in some places vs getting a fresh config with `connector.ServerIdentity.TLSConfig(cfg.CipherSuites)` in others. 

Given that `Clone` is a "shallow" clone, I suspect its safer to just create a fresh config for each separate step, unless we explicitly want the pointer-field-dependency of shallow clones.

**Review by espadolini — APPROVED** [review]

**Comment by nklaassen** [comment]
What is the problem with not returning the custom protocol? Just trying to understand if the way we use `teleport-proxy-grpc` is problematic too, it relies on the client requesting the protocol for routing, but IIRC we actually have to negotiate to `h2` for grpc to work

**Comment by espadolini** [comment]
> What is the problem with not returning the custom protocol?

It breaks the ALPN contract; the entire point of ALPN is that the client presents a list of protocols and the server chooses one that's supported _and makes the client aware_, without adding roundtrips to the communication.

If we're choosing to do `teleport-proxy-grpc` we should select that and have both sides aware of it. If that wouldn't work with clients in the wild today then we have to rethink the approach a bit I guess. Maybe a separate `teleport-proxy-grpc-but-this-time-we-actually-read-the-rfc` nextproto that both servers and client can ask and use?

**Comment by nklaassen** [comment]
> > What is the problem with not returning the custom protocol?
> 
> It breaks the ALPN contract; the entire point of ALPN is that the client presents a list of protocols and the server chooses one that's supported _and makes the client aware_, without adding roundtrips to the communication.
> 
> If we're choosing to do `teleport-proxy-grpc` we should select that and have both sides aware of it. If that wouldn't work with clients in the wild today then we have to rethink the approach a bit I guess. Maybe a separate `teleport-proxy-grpc-but-this-time-we-actually-read-the-rfc` nextproto that both servers and clients can ask and use?

I meant like, is there an actual problem/bug/situation where this breaks, that we need to fix or else things aren't going to work?

Not sure I'd go so far as to say it breaks the ALPN contract for the client to advertise support for `[teleport-proxy-grpc, h2]`, the server selects `h2`, and the actual protocol used between the two is `h2`. `teleport-proxy-grpc` is not a real protocol that's supported by any server, so I guess our client is lying a bit, but it is a very useful way to pass that routing information

File: lib/service/service.go (hunk 1 above)
Comment (nklaassen) [reply lib/service/service.go:4548]: Yeah this is what I was mentioning here https://github.com/gravitational/teleport/pull/43473#issuecomment-2192031355, IIRC the server has to actually select `h2` for the `google.golang.org/grpc` code to work

**Review by nklaassen — APPROVED** [review]

**Pull request merged**